### ISSUE TITLE
aria-readonly on input generates false positive

### ISSUE THREAD

**Issue by babielmam** [validator/validator#1199]
An input type="checkbox" with aria-readonly="true" generates a false positive:
```
<!DOCTYPE html>
<html>
<head>
  <title>test</title>
</head>
<body>
  <input type="checkbox" name="test" aria-readonly="true" value="yes">
</body>
</html>
```

```
aria-readonly not allowed on element input at this point.
```

The reason I use aria-readonly and not native readonly is, that the native attribute supposedly is only supported on text inputs. 

The HTML spec says in the [section for global attributes](https://html.spec.whatwg.org/multipage/dom.html#global-attributes):

> To enable assistive technology products to expose a more fine-grained interface than is otherwise possible with HTML elements and attributes, a set of annotations for assistive technology products can be specified (the ARIA role and aria-* attributes).

**Comment by sideshowbarker** [comment]
@scottaohara, @stevefaulkner is this a bug I should fix? Is the checker not conforming on this?

**Comment by babielmam** [comment]
I dug a little deeper into the ARIA spec and found further information. In fact the last sentence in my issue is a bit misleading as it is not entirely relevant, since it says, that other _global_ aria-attributes are not excluded. However, "aria-readonly" is not a global aria-attribute.

Nonetheless, input type="checkbox" implicitly has role="checkbox" and under certain circumstances can also have the role menuitemcheckbox, option or switch. In case it either has the role checkbox, menuitemcheckbox or switch, aria-readonly is explicitly listed as allowed attribute. See these sources:
https://www.w3.org/TR/wai-aria-1.1/#checkbox
https://www.w3.org/TR/wai-aria-1.1/#switch
https://www.w3.org/TR/wai-aria-1.1/#menuitemcheckbox

**Comment by sideshowbarker** [comment]
@babielmam Sounds like you’re right to me, but I’ll wait for confirmation from @scottaohara @stevefaulkner (who are intimately familiar with all this stuff)

**Comment by scottaohara** [comment]
This is a tough one, because the `readonly` attribute is not allowed on a native checkbox in HTML.  ARIA does allow the `aria-checked` attribute to be used on an element with `role=checkbox`, and a native checkbox input would implicitly expose that role.

Support for the attribute on a native checkbox is incredibly poor.  An element with an explicit `role=checkbox` and `aria-readonly` has a mildly better rate of exposure: [test results](https://codepen.io/scottohara/pen/KKmoGWy)

To be honest, I absolutely do not think that there should be an error thrown if the following markup were checked:
```html
<div role=checkbox aria-checked=true tabindex=0 aria-readonly=true>...</div>
```
Full support is lacking, but it's valid to the ARIA spec.

I continue find it hard to be supportive of `aria-readonly` on a native checkbox per how HTML has defined `readonly` and it not being applicable to a checkbox. The fact it presently has such poor support in exposing the state doesn't help either.

Finally, I generally don't understand how a 'readonly' checkbox could not be better communicated by using `aria-disabled` on the native checkbox instead, and then using JS to remove the ability to toggle the control?  I understand you might not want to use `disabled` on the native input if you have to submit the value associated with that checkbox, in a form.  But, using `aria-disabled` wouldn't have that 'side effect'.  It would be consistently communicated, and - in the case of a 'readonly checkbox' there'd be no real difference in functionality.  the checkbox could remain focusable, but you couldn't change it's current checked state.  


sorry this isn't as neat of a yes / no as one might have hoped.  but per the actual support for this attribute, i'm not sure if one should continue to want to even try using this pattern unless they are also willing to petition to browsers/AT to make sure the attribute is accurately exposed.

**Comment by babielmam** [comment]
I mostly agree with you @scottaohara. 
However I actually see a difference in a disabled checkbox and a read-only checkbox. I read about an example in another discussion, where there is a checkbox in an OS - it can be disabled, because the device doesn't support the option or it can be read-only, because the current user does not have high enough privilege to change the setting. That's a perfectly valid example for me to have a read-only checkbox. However, I also see this as a bug in the HTML spec and not in the ARIA spec, since it's allowed in ARIA.

You mention that an arbitrary element with explicit role=checkbox and the aria-readonly is valid, but lacks support. I don't see, how this can be more or less valid than an input with type=checkbox. 

In my opinion the validator should test conformance with the spec and not whether the solution is supported in a wide range of devices. I completely agree with your concerns about support and that is also why I will change our real-world solution to this problem. But this should not be of concern for the validator in my opinion.

**Comment by scottaohara** [comment]
I've had similar discussions with others about the distinctions between read-only vs disabled checkboxes.  The HTML spec is clear in their stance on this subject per the following quote that accompanies the `readonly` attribute's section in the spec:

>The difference between disabled and readonly is that read-only controls can still function, whereas disabled controls generally do not function as controls until they are enabled. ... Only text controls can be made read-only, since for other controls (such as checkboxes and buttons) there is no useful distinction between being read-only and being disabled, so the readonly attribute does not apply.

This is related to answering your next statement:

>You mention that an arbitrary element with explicit role=checkbox and the aria-readonly is valid, but lacks support. I don't see, how this can be more or less valid than an input with type=checkbox.

The distinction here simply comes down to what is allowed by each specification.  HTML doesn't allow checkboxes to be read-only.  ARIA does allow this... but ARIA is not meant to be used as a loop hole to force in semantics the host language has purposefully not implemented.  Hence my example being valid, because the semantics are entirely exposed via ARIA attributes and in conformance with that specification, and do not tread into the territory of trying to loophole what the native HTML checkbox doesn't allow.  In other words `<input type=checkbox>` is not the same as `<div role=checkbox>`, hence allowances for each can differ.  

Regarding your opinion on whether the validator should indicate support for features or not, I was not attempting to make a claim that it should.  I actually tend to agree with you on this.  That said, I think understanding the full picture of the ask is quite useful in helping to make decisions like this. Or put another way, if `aria-readonly` were to not raise a validation error on `<input type=checkbox>` who would that really help in the end, knowing the support for this is horrible?

**Comment by dd8** [comment]
VNU now reports

> Element input is missing required attribute aria-checked
> Element input is missing required attribute role

These are not reported if the aria-readonly attribute is removed - needs further investigation. This seems related to other issues like https://github.com/validator/validator/issues/1803 and https://github.com/validator/validator/issues/1714
